pkg/cmd/codespace/ssh.go

```diff
@@ -147,6 +161,146 @@ func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err e
 	}
 }
 
+func (a *App) printOpenSSHConfig(ctx context.Context, opts configOptions) error {
+	// Ensure all child tasks (e.g. port forwarding) terminate before return.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var err error
+	var codespaces []*api.Codespace
+	if opts.codespace == "" {
+		a.StartProgressIndicatorWithLabel("Fetching codespaces")
+		codespaces, err = a.apiClient.ListCodespaces(ctx, -1)
+		a.StopProgressIndicator()
+	} else {
+		var codespace *api.Codespace
+		codespace, err = getOrChooseCodespace(ctx, a.apiClient, opts.codespace)
+		codespaces = []*api.Codespace{codespace}
+	}
+	if err != nil {
+		return fmt.Errorf("error getting codespace info: %w", err)
+	}
+
+	t, err := template.New("ssh_config").Parse(`Host cs.{{.Name}}.{{.EscapedRef}}
+	User {{.SSHUser}}
+	ProxyCommand {{.GHExec}} cs ssh -c {{.Name}} --stdio
+	UserKnownHostsFile=/dev/null
+	StrictHostKeyChecking no
+	LogLevel quiet
+	ControlMaster auto
+
+`)
+	if err != nil {
+		return fmt.Errorf("error formatting template: %w", err)
+	}
+
+	ghexec, err := os.Executable()
+	if err != nil {
+		return err
+	}
+
+	// store a mapping of repository -> remote ssh username. This is
```

Comment:
How is this mapping not degenerate? In other words, if the ssh username may vary, what's to stop it from varying even across multiple codespaces for the same repository? I would expect the correct key here is the codespace identity (at which point there's no benefit to the map).

Reply:
That's a good question - I'm assuming that a given repository always uses the same username here. If it can vary (and I suppose it can if different branches use different devcontainers? I'm not certain), then this caching isn't correct.

Ideally, we'd refactor the backend to separate fetching the ssh username and actually running sshd. That would let us remove most of this complexity, and generate configs for all codespaces (available or not).

Reply:
Except in rare instances where a codespace somehow uses multiple usernames, though, I think this is a good thing. If a user has many codespaces, I expect it's common that a lot of them are different instances of the same repo, in which case this is a sizable speedup.

Reply:
This is moot after https://github.com/cli/cli/pull/4915/commits/06eb5ad47fbb055046bb3c86ee12587d7a714c93